File: sdk/python/greenlab/batch.py
"""Batch building and td-offset calculation for the GreenLab SDK."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import List, Optional

import msgpack

from .schema import ChannelSchema

MAX_TD_MILLIS = 65535


@dataclass
class Reading:
    field_name: str
    value: float
    ts: datetime


def build_batches(
    readings: List[Reading],
    schema: ChannelSchema,
    schema_version: int,
) -> List[bytes]:
    """Partition *readings* into one or more MessagePack-encoded batch payloads.

    Readings within a batch share a base timestamp (the timestamp of the first
    reading in the batch). Per-field ``td`` values are millisecond offsets from
    the base. If any delta exceeds ``MAX_TD_MILLIS`` (65535ms), the current
    batch is flushed and a new one begins at the offending reading's timestamp.

    Returns a list of raw MessagePack bytes, one element per batch.
    Raises :class:`ValueError` for empty input or unknown field names.
    """
    if not readings:
        raise ValueError("sdk: no readings to send")

    num_fields = len(schema.ordered_fields)
    if num_fields == 0:
        raise ValueError("sdk: schema has no fields")

    # Build fieldName → 0-based position in the f array.
    field_pos = {fe.name: i for i, fe in enumerate(schema.ordered_fields)}

    for r in readings:
        if r.field_name not in field_pos:
            raise ValueError(f"sdk: unknown field {r.field_name!r}")

    result: List[bytes] = []
    base_ts: Optional[datetime] = None
    f_vals: List[Optional[float]] = [None] * num_fields
    td_vals: List[Optional[int]] = [None] * num_fields
    batch_used = False

    def flush() -> None:
        nonlocal batch_used
        # Only include slots with values.
        f: List[float] = []
        td: List[int] = []
        for i in range(num_fields):
            if f_vals[i] is not None:
                f.append(f_vals[i])  # type: ignore[arg-type]
                td.append(td_vals[i])  # type: ignore[arg-type]
        if not f:
            return
        assert base_ts is not None
        payload = {
            "sv": schema_version,
            "ts": int(base_ts.timestamp()),
            "f": f,
            "td": td,
        }
        result.append(msgpack.packb(payload, use_bin_type=True))
        batch_used = False

    def reset(new_base: datetime) -> None:
        nonlocal base_ts, f_vals, td_vals, batch_used
        base_ts = new_base
        f_vals = [None] * num_fields
        td_vals = [None] * num_fields
        batch_used = False

    reset(readings[0].ts)

    for r in readings:
        assert base_ts is not None
        delta_ms = int((r.ts - base_ts).total_seconds() * 1000)
        if delta_ms < 0:
            delta_ms = 0

        if delta_ms > MAX_TD_MILLIS:
            if batch_used:
                flush()
            reset(r.ts)
            delta_ms = 0

        pos = field_pos[r.field_name]
        f_vals[pos] = r.value
        td_vals[pos] = delta_ms
        batch_used = True

    if batch_used:
        flush()

    return result

File: sdk/go/batch.go
package sdk

import (
	"fmt"
	"time"

	"github.com/vmihailenco/msgpack/v5"
)

const maxTDMillis = 65535

// reading represents a single field measurement at a point in time.
type reading struct {
	fieldName string
	value     float64
	ts        time.Time
}

// msgpackBatch is the wire-format struct for a batch payload.
// Field names use compact short keys matching the server's compact format.
type msgpackBatch struct {
	SV uint32    `msgpack:"sv"`
	TS int64     `msgpack:"ts"`
	F  []float64 `msgpack:"f"`
	TD []uint16  `msgpack:"td"`
}

// buildBatches partitions readings into one or more msgpackBatch values.
//
// Readings are grouped by their schema-ordered field position. The base
// timestamp is the timestamp of the first reading in each group. If any
// reading's delta from the base exceeds 65535ms, a new batch is started
// at that point.
//
// The returned [][]byte contains the MessagePack-encoded payload for each
// batch, ready to be optionally compressed and sent.
func buildBatches(readings []reading, schema channelSchema, schemaVersion uint32) ([][]byte, error) {
	if len(readings) == 0 {
		return nil, fmt.Errorf("sdk: no readings to send")
	}

	// Determine the number of fields from the schema.
	numFields := len(schema.orderedFields)
	if numFields == 0 {
		return nil, fmt.Errorf("sdk: schema has no fields")
	}

	// Build a lookup: fieldName → position in orderedFields (0-based).
	// The "f" array is positional, matching the server's ordered field list.
	fieldPos := make(map[string]int, numFields)
	for i, fe := range schema.orderedFields {
		fieldPos[fe.Name] = i
	}

	// Validate that all readings map to known fields.
	for _, r := range readings {
		if _, ok := fieldPos[r.fieldName]; !ok {
			return nil, fmt.Errorf("sdk: unknown field %q", r.fieldName)
		}
	}

	var (
		result    [][]byte
		baseTS    = readings[0].ts
		fVals     = make([]float64, numFields)
		tdVals    = make([]uint16, numFields)
		hasValue  = make([]bool, numFields)
		batchUsed bool
	)

	flushBatch := func() error {
		// Only include slots that have values.
		f := make([]float64, 0, numFields)
		td := make([]uint16, 0, numFields)
		for i := 0; i < numFields; i++ {
			if hasValue[i] {
				f = append(f, fVals[i])
				td = append(td, tdVals[i])
			}
		}
		if len(f) == 0 {
			return nil
		}
		batch := msgpackBatch{
			SV: schemaVersion,
			TS: baseTS.Unix(),
			F:  f,
			TD: td,
		}
		encoded, err := msgpack.Marshal(batch)
		if err != nil {
			return fmt.Errorf("sdk: encode batch: %w", err)
		}
		result = append(result, encoded)
		return nil
	}

	resetBatch := func(newBase time.Time) {
		baseTS = newBase
		fVals = make([]float64, numFields)
		tdVals = make([]uint16, numFields)
		hasValue = make([]bool, numFields)
		batchUsed = false
	}

	for _, r := range readings {
		deltaMs := r.ts.Sub(baseTS).Milliseconds()
		if deltaMs < 0 {
			deltaMs = 0
		}

		// If delta exceeds uint16 range, flush current batch and start a new one.
		if deltaMs > maxTDMillis {
			if batchUsed {
				if err := flushBatch(); err != nil {
					return nil, err
				}
			}
			resetBatch(r.ts)
			deltaMs = 0
		}

		pos := fieldPos[r.fieldName]
		fVals[pos] = r.value
		//nolint:gosec // deltaMs is bounded by maxTDMillis = 65535 which fits uint16
		tdVals[pos] = uint16(deltaMs)
		hasValue[pos] = true
		batchUsed = true
	}

	// Flush remaining readings.
	if batchUsed {
		if err := flushBatch(); err != nil {
			return nil, err
		}
	}

	return result, nil
}
